enhancements/ingress/global-options-enable-hsts.md

```diff
@@ -0,0 +1,411 @@
+
+With the addition of processing for `ROUTER_DEFAULT_HSTS`, the code template snippet above becomes:
+```gotemplate
+{{- $router_global_hsts := env "ROUTER_DEFAULT_HSTS" "" }}
+...
+{{- if matchValues (print $cfg.TLSTermination) "edge" "reencrypt" }}
+  {{- /* First check the environment to see if the header is set globally */}}
+  {{- if (ne "" $router_global_hsts) }} 
+    http-response set-header Strict-Transport-Security {{$router_global_hsts}}
+  {{- end }}{{/* global_hsts */}}
+  {{- /* Override if the annotation exists.  Set-header replaces any existing header */ }}
+  {{- with $hsts := firstMatch $hstsPattern (index $cfg.Annotations "haproxy.router.openshift.io/hsts_header") }}
+    http-response set-header Strict-Transport-Security {{$hsts}}
+  {{- end }}{{/* hsts header override per-route */}}
+{{- end }}{{/* is "edge" or "reencrypt" */}}
+```
+
+#### Enable HSTS per-router
+Some of the proposal for this goal is in the section above.  In order to allow the HAProxy template to add HSTS Headers to
+a limited list of routers, it will use the previously mentioned field `InScopeRouters` in the `IngressControllerHSTSHeaderPolicy`
+type.
+
+The details on the processing for limited scope in `desiredRouterDeployment` are completed here.
+Add one more environment variable, `ROUTER_HSTS_LIST`, which will contain the list of domains in `InScopeRouters`:
+```go
+	const RouterHSTSHeaderList        = "ROUTER_HSTS_LIST"
+
+       if scope == "limited" {
+            env = append(env, corev1.EnvVar{Name: RouterHSTSHeaderScope, Value: "limited"})
+            env = append(env, corev1.EnvVar{Name: RouterHSTSHeaderList, Value: ci.Spec.HTTPHeaders.HSTSHeaderPolicy.InScopeRouters})
+       }
+```
+As shown in this example, the HSTS Policy scope is for a limited list of two routers, expires after one year, includes
+subdomains, and requests preload:
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: IngressController
+metadata:
+  name: default
+  namespace: openshift-ingress-operator
+spec:
+  httpHeaders:
+    hstsHeaderPolicy:
+      scope: all
+      inScopeRouters:
+        - domainA
+        - domainB
+      max-age: 31536000
+      includeSubDomains: true
+      preload: true
+```
+
+With the addition of processing for `ROUTER_HSTS_SCOPE` of `limited`, and `ROUTER_HSTS_LIST`, the code template snippet
+above becomes:
+```gotemplate
+{{- $router_hsts := env "ROUTER_DEFAULT_HSTS" "" }}
+{{- $router_hsts_scope := env "ROUTER_HSTS_SCOPE" "" }}
+{{- $router_hsts_list := env "ROUTER_HSTS_LIST" }}
+...
+{{- if matchValues (print $cfg.TLSTermination) "edge" "reencrypt" }}
+  {{- /* First check the environment to see if the header is set globally */}}
+  {{- if (ne "" $router_hsts) }}
+    {{- /* Now check the environment to see if scope is all, or limited but request domain is in the limited list */}}
+    {{- if or (eq "all" $router_hsts_scope) (and (eq "limited" $router_hsts_scope) (eq true (InList req.hdr(Host) $router_hsts_list)) )) }}
```

Comment:
Everything inside the double-braces (`{{ ... }}`) is Go template, and `req.hdr` is an HAProxy configuration directive, so this will not work properly.  It may be necessary to write out a map file and define an acl (`acl hsts req.hdr(host),lower,map(/var/lib/haproxy/conf/os_hsts.map) -m found`) to use below (`http-response set-header Strict-Transport-Security {{$router_hsts}} if hsts`).

Reply:
I reworked this.